latch/registry/utils.py

```diff
@@ -0,0 +1,343 @@
+        except KeyError:
+            raise ValueError(
+                "cannot convert to python - malformed registry enum type without"
+                " members"
+            )
+
+        if clean(value) not in members:
+            raise ValueError(
+                f"unable to convert {value} ({clean(value)}) to any of the enum members"
+                f" {members}"
+            )
+
+        return to_python_type(registry_type)._member_map_.get(clean(value))
+
+    if primitive == "blob":
+        typ = LatchFile
+        if (
+            "metadata" in registry_type
+            and "nodeType" in registry_type["metadata"]
+            and registry_type["metadata"]["nodeType"] == "dir"
+        ):
+            typ = LatchDir
+
+        if not (type(value) is dict and ("ldataNodeId" in value)):
+            raise ValueError("cannot convert non-blob value to blob type")
+
+        path_data = execute(
+            """
+            query nodePath($nodeId: BigInt!) {
+                ldataOwner(argNodeId: $nodeId)
+                ldataGetPath(argNodeId: $nodeId)
+            }
+            """,
+            {"nodeId": value["ldataNodeId"]},
+        )
+
+        url = path_data["ldataGetPath"]
+        owner = path_data["ldataOwner"]
+
+        if url == None:
+            raise ValueError("unable to convert to LatchFile/Dir as no url was found")
+
+        account_match = account_root_regex.match(url)
+        mount_match = mount_regex.match(url)
+
+        path = ""
+
+        if account_match is not None:
+            if owner is None:
+                raise ValueError(
+                    "unable to convert to LatchFile/Dir as owner was not found"
+                )
+
+            path = f"latch://{owner}.account{account_match.groupdict()['path']}"
+        elif mount_match is not None:
+            groups = mount_match.groupdict()
+            path = f"latch://{groups['mount']}.mount{groups['path']}"
+        else:
+            raise ValueError(
+                f"unable to convert to LatchFile/Dir as url was malformed: {url}"
+            )
+
+        return typ(path)
+
+    if primitive == "link":
+        if "sampleId" not in value:
+            raise ValueError("unable to convert to Link as row id was not found")
+        try:
+            table_id = registry_type["experimentId"]
+        except KeyError:
+            raise ValueError("Unable to convert link type without a table reference")
+
+        from latch.registry.table import Table
+
+        tbl = Table(table_id)
+
+        for page in tbl.list_records():
+            for record in page["records"]:
+                if record.id == value["sampleId"]:
+                    return record
+
+        raise ValueError(
+            "unable to convert registry link to python - cannot find corresponding row"
+            " in linked table"
+        )
+
+    python_type = to_python_type(registry_type)
+    if isinstance(value, python_type):
```

Comment:
I'd deal with primitives individually, catalog is quite limited in what it supports + you would ideally validate that your `python_type` matches the catalog type